### ISSUE TITLE
[BUG] Module resources don't seem to load when navigating to a page with the module

### ISSUE THREAD

**Issue by iJungleboy** [oqtane/oqtane.framework#4342]
### Oqtane Info

Version - 5.1.2
Render Mode - Static 
Database - SQL Server

### Describe the bug

A module can describe resources to load - JS/CSS.
The idea is that these are automatically added to the page by Oqtane.

But on our tests, it only appears to work if the first full load happens on the page with the module. 

### Steps To Reproduce

So in our tests, this works:

1. module on page About-Us
2. Fully load page About us
3. all is ok

...but this fails:

1. Module is on page About us
2. fully load page home
3. browse to About us

Resources seem to not load. 

### Anything else?

The module is configured to have `RenderMode.Static`.

@tvatavuk just fyi

**Comment by Raceeend** [comment]
I noticed that OnInitializedAsync was not always called when the Site is set to Interactive. The setting of the module then has no influence on this. If the Site setting is set to Static, OnInitializedAsync will be called every time.

Maybe you can test by setting the Site (UI Component Settins) to Static.

**Comment by zyhfish** [comment]
If this is for js files, you can make your js script to have below functions:
```
export function onLoad() {
...
}

export function onUpdate() {
...
}

export function onDispose() {
...
}
```

then register your js file with below format:
```
new Resource { ResourceType = ResourceType.Script, Url = "[Js File URL]", Location = ResourceLocation.Body, Reload = true, RenderMode = RenderModes.Static }
```

in this case your code will be executed every time navigate to the page.

**Comment by iJungleboy** [comment]
@zyhfish that's good for code that must run every time, but not for external files that should be loaded - especially larger code.

**Comment by sbwalker** [comment]
@iJungleboy it is not clear if you are saying that the JavaScript references are not being added to the DOM, or if they are not being executed (ie. calling onload), or both. 

The fact that Blazor does not call onload when navigating from page to page has been well documented previously - it is due to Blazor's "enhanced navigation" which I covered in the "building beautiful websites" webinar I presented 2 months ago (based on the Arsha theme). In scenarios where you need JavaScript to execute on navigation you need to hook the onupdated event. 

Can you please provide more information on your specific scenario.

**Comment by iJungleboy** [comment]
@sbwalker thx for looking into this. The module has

```
    public override List<Resource> Resources => new()
    {
        new Resource { ResourceType = ResourceType.Script, Url = $"Modules/{OqtConstants.PackageName}/Module.js" }
    };
```

If it's on the initially loaded page - eg. ctrl-F5 - it's available.

If the user starts on another page, and hen navigates to the page with the module, it's not.

**Comment by sbwalker** [comment]
@iJungleboy you did not actually answer my question. I was asking if you viewed the source code for the page in your browser dev tools and verified if the link element reference to Module.js is in the page head or not?

**Comment by tvatavuk** [comment]
Thanks to everyone for helping!

To test and understand the new rendering modes and their behaviors, we are using the Oqtane 5.1.2 source code (master). Our setup includes Windows 11, Visual Studio 2022 v17.10.3, and latest .NET 8.0 SDK with all the latest updates.

We've created several sub-portals for different tests. Specifically, to test Static SSR behaviors with the default Oqtane site template and theme, we created a new sub-site. The home page is accessible at http://localhost:44357/ss10 and it is defult one, without custom module. On the page http://localhost:44357/ss10/t1, we developed a new custom Oqtane module, ToSic.Module.Test512, using the default Oqtane module template.

Here are our observations:

1. Module.js in Resources: Yes, the custom Oqtane module, ToSic.Module.Test512, includes Module.js in its resources.

![image](https://github.com/oqtane/oqtane.framework/assets/2724850/847b412f-0147-438c-a31f-46301973481f)

2. Initial Page Visit: Yes, the first page visited did not contain the module.

3. Navigation to Module Page: Yes, after home we are navigating to a page with the module.

![image](https://github.com/oqtane/oqtane.framework/assets/2724850/228ce3c5-f1b2-492c-976b-dfa91a990501)

4. Module.js appeared in the source code, but Module.js is not Loading.

![image](https://github.com/oqtane/oqtane.framework/assets/2724850/5c9b228d-a855-4872-b6f9-cbecb11b6c75)


I will provide more details in the next comment.

**Comment by tvatavuk** [comment]
To better support different cases of loading resources in Static SSR, we created a simple service to distinguish between the first full page load on Static SSR and subsequent page loads that occur with Blazor enhanced navigation.

Here is the helper service implementation:

## Interface Definition
```cs
 namespace ToSic.Sxc.Oqt.Shared.Interfaces
{
    /// <summary>
    /// Service for providing rendering information based on the current HTTP context and module render mode.
    /// </summary>
    public interface IRenderInfoService
    {

        /// <summary>
        /// Checks if the module render mode is static SSR (Server-Side Rendering).
        /// </summary>
        /// <param name="renderMode">The module render mode.</param>
        /// <returns>True if the render mode is static SSR, otherwise false.</returns>
        bool IsStaticSsr(string renderMode);

        /// <summary>
        /// Checks if Blazor Enhanced Navigation is enabled for static SSR.
        /// </summary>
        /// <param name="renderMode">The module render mode.</param>
        /// <returns>True if Blazor Enhanced Navigation is enabled, otherwise false.</returns>
        /// <remarks>
        /// Blazor Enhanced Navigation is a feature in .NET 8 that allows for progressively-enhanced navigation in multi-page apps.
        /// This is enabled when the app loads `blazor.web.js` and does not use an interactive Router. It works by intercepting 
        /// navigation within the base href URI space, loading content via `fetch` requests, and syncing the DOM.
        /// </remarks>
        bool IsBlazorEnhancedNav(string renderMode);

        /// <summary>
        /// Checks if the SSR Framing response header exists, indicating a fetch page request in Blazor Enhanced Navigation.
        /// </summary>
        /// <param name="renderMode">The module render mode.</param>
        /// <returns>
        /// True if the SSR Framing response header exists (indicating a fetch page request during Blazor Enhanced Navigation).
        /// False if it is a standard full page load in the browser.
        /// </returns>
        /// <remarks>
        /// This helps distinguish between a full page load (initial load) and a fetch page request (subsequent navigation within the app).
        /// This behavior occurs when Blazor Enhanced Navigation is in use for static SSR.
        /// </remarks>
        bool IsSsrFraming(string renderMode);
    }
}
```

## Server-Side Implementation
```cs
using Microsoft.AspNetCore.Http;
using Oqtane.Shared;
using System;
using System.Collections.Generic;
using System.Linq;
using ToSic.Sxc.Oqt.Shared.Interfaces;

namespace ToSic.Sxc.Oqt.Server.Services
{
    /// <inheritdoc />
    public class RenderInfoService(IHttpContextAccessor httpContextAccessor) : IRenderInfoService
    {
        private readonly List<string> _enhancedNavValues = ["allow", "on"];
        private const string BlazorEnhancedNav = "blazor-enhanced-nav";
        private const string SsrFraming = "ssr-framing";

        /// <inheritdoc />
        public bool IsStaticSsr(string renderMode = RenderModes.Interactive)
          => string.Equals($"{renderMode}", RenderModes.Static, StringComparison.OrdinalIgnoreCase);

        /// <inheritdoc />
        public bool IsBlazorEnhancedNav(string renderMode)
        {
            if (!IsStaticSsr(renderMode))
                return false;

            var context = httpContextAccessor.HttpContext;
            if (context != null && context.Response.Headers.TryGetValue(BlazorEnhancedNav, out var enhancedNav))
                return _enhancedNavValues.Contains($"{enhancedNav}", StringComparer.OrdinalIgnoreCase);

            return false;
        }

        /// <inheritdoc />
        public bool IsSsrFraming(string renderMode)
        {
            if (!IsBlazorEnhancedNav(renderMode))
                return false;

            var context = httpContextAccessor.HttpContext;
            return context != null && context.Response.Headers.TryGetValue(SsrFraming, out _);
        }
    }
}

```

## Client-Side Implementation
```cs
using ToSic.Sxc.Oqt.Shared.Interfaces;

namespace ToSic.Sxc.Oqt.Client.Services
{
    /// <inheritdoc />
    public class RenderInfoService : IRenderInfoService
    {
        /// <inheritdoc />
        public bool IsStaticSsr(string renderMode) => false;

        /// <inheritdoc />
        public bool IsBlazorEnhancedNav(string renderMode) => false;

        /// <inheritdoc />
        public bool IsSsrFraming(string renderMode) => false;
    }
}

```

This service helps distinguish between a full page load and a fetch page request during Blazor Enhanced Navigation. This way, you can handle resource loading more appropriately based on the context of the page load.

**Comment by tvatavuk** [comment]
To ensure loading of external .js resource like one in example we used solution based on **turnOn** that was invented by @iJungleboy.

Example of external resource that we loaded with **turnOn**.

```html
<script integrity="sha256-0af2VbC4vmPsa8OLBAKBmLoyuKq4bBlKK2KOgMWayio=" crossorigin="anonymous" src="https://cdn.jsdelivr.net/npm/@fancyapps/ui@4.0.31/dist/fancybox.umd.js"></script>
```

More info about **turnOn**: https://docs.2sxc.org/js-code/turn-on/index.html

**Comment by iJungleboy** [comment]
just to clarify what @tvatavuk said - **turnOn** doesn't ensure that the JS is loaded into the browser, it's job is to determine when all the requirements of a script are given, and then start some code. 

A bit like a browser onLoad, but in a way that can handle late-added scripts, complicated dependencies, and CSP safe since it's configuration only, meaning that it can be used even in very strict CSP scenarios. CSP = Client Security Policy.

**Comment by sbwalker** [comment]
@tvatavuk thank you for confirming that Oqtane is injecting the reference to the JavaScript resource into the page DOM - this means that Oqtane is behaving properly. The fact that the JavaScript was not being executed is not an Oqtane issue - it is related to Blazor SSR and how Enhanced Navigation behaves in .NET 8. There is plenty of information available online about this behavior (ie. https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/static-server-rendering?view=aspnetcore-8.0). As explained in the Building Beautiful Websites webinar, Oqtane is utilizing a solution created by Mackinnon Buck from the Blazor Dev Team which enables JavaScript navigation callbacks (https://github.com/MackinnonBuck/blazor-page-script).  If this solution does not work for you, you are free to use your own custom solution.

**Comment by sbwalker** [comment]
@tvatavuk you mentioned that the `page-script` magic doesn't work with some external .js resources, as illustrated in this example:

```html
<script integrity="sha256-0af2VbC4vmPsa8OLBAKBmLoyuKq4bBlKK2KOgMWayio=" crossorigin="anonymous" src="https://cdn.jsdelivr.net/npm/@fancyapps/ui@4.0.31/dist/fancybox.umd.js"></script>
```

I assume this is because the page-script does not support integrity or crossorigin attributes. Would it make sense for me to reach out to Mackinnon Buck to provide this feedback and explore a potential solution? A lot of developers are using the page-script project as it is the official example provided by Microsoft, so it would benefit the community to address this scenario.

Page-script (https://github.com/MackinnonBuck/blazor-page-script) already provides the ability to distinguish between the first full page load on Static SSR and subsequent page loads that occur with Blazor enhanced navigation. It exposes different events for these purposes - onLoad and onUpdate. It would be useful to familiarize yourself with the capabilities as I would assume the page-script approach will eventually become the "standard" approach in Blazor for dealing with JavaScript in SSR Enhanced Navigation.

**Comment by tvatavuk** [comment]
Thank you for bringing up the issue with the `page-script` magic and external JavaScript resources. Let me clarify the situation.

The problem arises because the external JavaScript module we're using, https://cdn.jsdelivr.net/npm/@fancyapps/ui@4.0.31/dist/fancybox.umd.js, is in the classic  [UMD (Universal Module Definition)](https://github.com/umdjs/umd) format, whereas `page-script` magic relies on the standard dynamic import, which expects modules in the [ESM (ECMAScript module)](https://v8.dev/features/modules) format. Due to this format discrepancy, the `page-script` magic does not work as expected.

A potential solution would be to use an external JavaScript module in the ESM format, such as https://cdn.jsdelivr.net/npm/@fancyapps/ui@4.0.31/dist/fancybox.esm.js. However, this requires changes to the existing code in production, which might not be feasible immediately.

Regarding the inclusion of additional script attributes like `integrity` and `crossorigin`, more investigation is necessary. These attributes enhance security of the external scripts in browser. However, their interaction with the dynamic `import()` functionality, which already has specific handling for CORS, needs to be understood better. If @iJungleboy or anyone, has more experience or insights on this matter, suggestions would be greatly appreciated.

Thanks for explaining Page-script (https://github.com/MackinnonBuck/blazor-page-script) and its ability to distinguish between the first full page load on Static SSR and subsequent page loads that occur with Blazor enhanced navigation. Yes, I will familiarize myself with these capabilities and find how to use them in 2sxc. In our case we needed info about Static SSR and enhanced navigation during razor rendering that happens on server to provide correct list of resources (js, css) to browser.

**Comment by iJungleboy** [comment]
@sbwalker thanks for your help and support 🙏🏼

**Comment by sbwalker** [comment]
@tvatavuk I reached out to Mackinnon Buck and he explained that that page-script custom element can include additional attributes - he created an example here:

https://github.com/MackinnonBuck/blazor-page-script/commit/44ad6833b8375a9eaa9b0fa9ace70e024b823f2d

The most important part is that the attribute rel="modulepreload" needs to be included so that the dynamic module imports will respect the attributes set on the link element:

https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/modulepreload

Oqtane will need to be modified to include the additional attributes - and then it will be able to support remote resources.

**Comment by sbwalker** [comment]
One other thing I want to explore is if it is really necessary to force developers to include the Reload = True property in their Resources declaration. Instead, the framework could determine that the site is using Static rendering and use the page-script approach by default for Script resources. This would automatically wire up the onLoad, onUpdate, onDispose events if they exist - and if a JavaScript library does not utilize them, it should not cause any issues. This would simplify the experience for developers as they would not need to worry about an extra property which is only applicable for a specific render mode.

**Comment by mdmontesinos** [comment]
> One other thing I want to explore is if it is really necessary to force developers to include the Reload = True property in their Resources declaration. Instead, the framework could determine that the site is using Static rendering and use the page-script approach by default for Script resources. This would automatically wire up the onLoad, onUpdate, onDispose events if they exist - and if a JavaScript library does not utilize them, it should not cause any issues. This would simplify the experience for developers as they would not need to worry about setting an extra property which is only applicable for a specific render mode.

This would actually be a really nice addition for the developer, as I was also getting a bit confused with javascript in static ssr (see #4038). Although you would still need to implement the onLoad, onUpdate and onDispose events when required, so the problem does not totally disappear.

**Comment by tvatavuk** [comment]
Absolutely, I agree! 🎉 Simplifying the developer experience by having the framework automatically determine the need for `Reload = True` based on the rendering mode is a great idea. 🚀 This enhancement will undoubtedly reduce the complexity of custom module development in Oqtane. Thank you 👏

**Comment by iJungleboy** [comment]
After some more research I think we really need to clarify what `Reload = true` should mean. 
@sbwalker I'll also document it, as soon as it's clear to me 😉

Basically my interpretation is that `Reload = true` should mean 

1. it should get loaded **anew** on every "load" (whatever that means)
2. ...and that it should be re-executed (not just loaded) on every "load"

This would still pose the question:

1. load of _page_? <-- probably this
3. load of _module_ - e.g. what should happen, if the module is on the page multiple times?

Because I assume that's what `Reload = true` was originally meant to say.

ATM we're forced to use it to ensure loading of our scripts when soft-navigating to a page with the module on it - which is probably not the real intent of this.
I assume that a module can expect it's resources to always be loaded by Oqtane - no matter how the page was accessed. 
...which in turn means that the behavior in 5.1.2 where we _must_ provide `Reload=true` is a bug.

But since it's not fully clear what the official meaning `Reload=true` is, it's hard to create a bug report. 

@sbwalker I would appreciate clarification on this, so I can document it ;)

**Comment by sbwalker** [comment]
The Reload parameter was discussed in March in this thread: https://github.com/oqtane/oqtane.framework/issues/4038 when it was originally introduced.

The Reload parameter is only applicable when a module component is using Static Render Mode and it is only applicable to JavaScript resources. In this scenario, Blazor is using Enhanced Navigation which means that when you navigate to a new page, a full page refresh is not occurring - Blazor is patching the DOM based on the differences in the output (ie. it is behaving like a SPA). This means that ```<script>``` references are not refreshed in the DOM - which means they will not be reloaded by the browser. The Reload parameter instructs the framework that a JavaScript resource needs to be reloaded on page navigations.

**Comment by sbwalker** [comment]
@iJungleboy just to clarify so that I am able to reproduce the behavior you are describing:

- Site using Static Rendering
- Module which uses Static Rendering for its main component view is on page About Us
- fully load page Home
- browse to About Us page
- Resources for Module on About Us page do not load
- when you say "Resources" you are referring to a JavaScript resource which does NOT have Reload = True specified
- when you say "load" you mean that the ```<script>``` element for the Resource is NOT added to the page DOM

correct?

**Comment by iJungleboy** [comment]
Almost correct, but not quite. the last line:

> when you say "load" you mean that the <script> element for the Resource is NOT added to the page DOM

is wrong.

It is added, but of course nothing happens - the addition has no effect and is useless. 

The module creator specified this as a resources and expects the resource to be loaded - not added as a "comment" to the source code. 

What I'm saying is that on such a soft-reload, the `<script>`-addition should not happen, but instead be loaded using the loadJs.

**Comment by sbwalker** [comment]
@iJungleboy as I explained earlier in this thread, the fact that the script element is included in the DOM but is not executed on page navigations is exactly how Enhanced Navigation behaves in .NET 8 - it is explained in this Microsoft article:  https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/static-server-rendering?view=aspnetcore-8.0. The article also explains that the recommended solution for handling this scenario in Blazor applications is to use event listeners (ie. onLoad, onUpdate, onDispose). The page-script custom element concept created by Mackinnon Buck simplifies the ability to register event listeners. Oqtane has integrated the page-script custom element solution and it is invoked by setting Reload = True on your Script resources. 

So basically the problem you are reporting is already known, is a well-known behavior of Blazor, and a solution already exists in Oqtane (based on Microsoft's recommended approach). The page-script solution in the framework does not support external resources currently (ie. integrity, crossorigin) however Mackinnon Buck provided an example on how to support this which was mentioned earlier in this thread (https://github.com/MackinnonBuck/blazor-page-script/commit/44ad6833b8375a9eaa9b0fa9ace70e024b823f2d) and it will be integrated for the next Oqtane release.

In order to consider an alternate approach such as using LoadJS - we would need to understand why the recommended approach from Microsoft (ie. the page-script custom element) is not acceptable.

**Comment by iJungleboy** [comment]
@sbwalker I think there are ca. 3 different points here

1. the term `Reload = true` implies that it will be loaded again and again. This is a completely different need than "Load reliably" and in some cases this may be desired, but in others a reload again and again will itself cause new problems. LoadJS would be an example of a script which - if reloaded again and again - would result in very strange / unexpected behavior, since new copies will kind of replace previous ones, but the old ones will still run in memory - with an incorrect list of previous actions. 

2. Reloading the scripts again and again will have some very special challenges. For example (I haven't tested it) - I believe that this model will result in the browser taking up more and more memory and CPU with each page visit - potentially breaking down on a mobile device after 20-30 pages viewed. Again because it's a big difference if a JS is freshly loaded, often-reloaded, often-reinitialized etc. Some scripts will be fine, others will quickly take down the browser. 

3. The `blazor-page-script` is actually an [ESM](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) loader, not a "normal" JavaScript loader. There are a lot of differences to what actually happens - making it incompatible with various scripts out in the wild. So in some cases it's the better choice, in other cases it's the wrong choice. The way I see it, this is a solution for scripts which were specifically developed / tested / optimized for Oqtane - so "your own scripts" but not ideal (or incompatible) for existing stuff. This include standard libraries taken from CDNs and more. 

I personally admire the blazor-page-script idea. I could imagine using it to create a `<script-load>` tag which has he same idea, but uses real js loading for the "normal" case - though I think loadJs can also handle ESM modules. This should reliably load JS the "normal" way to make it less stressful for developers as to what's happening, and ensure that this `<script-load>` tag would probably have all the attributes a real script-tag can have....

And then I would ensure that it doesn't reload every single time - since I'm concerned that this will bite us hard in production. 

So my goal is to accept that Blazor has these challenges (for now), but make sure that Oqtane takes all the necessary steps that the module developer doesn't have to learn all this to get things to work - because Oqtane magically makes everything behave as would be expected intuitively. I'm quite sure we can do that 😉

**Comment by iJungleboy** [comment]
...but I would like to point out that we on our side (2sxc) found enough workarounds to get things going. So I'm really just trying to ensure a smooth developer experience for everybody else. 

I felt it important to mention this - since the discussion is getting longer and longer, and I know your resources are just as restricted as mine. So if you want to close this for now, I'm ok with this.

**Comment by sbwalker** [comment]
@iJungleboy you mentioned that you are using a workaround which involves calling the existing includeScripts() function that is part of interop.js. Can you please provide more context for how you are doing this in Static rendering.

**Comment by sbwalker** [comment]
or perhaps @tvatavuk can explain your workaround for calling the existing includeScripts() function that is part of interop.js?

**Comment by iJungleboy** [comment]
I'm almost away for my long leave, but some quick infos.

About three years ago as we were encouraging people to 

1. move away from jQuery
2. embrace CSP (Content Security Policy)
3. create 2sxc Apps which could run on Dnn and Oqtane

We needed a reliable system to **run** javascript despite all kinds of problems, such as

1. dependencies being loaded
2. HTML being loaded (in Oqtane that was tricky)
3. JavaScripts needing values which were generated to razor, but adding js like `var mid = @Module.Id;` would really confuse people ... and would be hard to get CSP signed, since the fingerprint would change for every output.

So we invented **turnOn** https://github.com/2sic/turnOn - explained better here https://go.2sxc.org/turnon

It's basically an html tag (similar to https://github.com/MackinnonBuck/blazor-page-script idea, except that it's not meant for loading JS, but for **running** JS when all requirements have been met, and passing parameters to the JS which were generated in the page.

A simple example would be this:

```js
<!-- super simple example without any parameters, will wait for window.myLib to exist -->
<turn-on turn-on='{ "run": "window.myLib.Run()" }' />

<!-- example with args, mainly good when calling existing js functions -->
<turn-on turn-on='{ "run": "window.alert()", "args": ["window.simple is now ready"] }' />

<!-- example with additional await and data-parameters -->
<turnOn turn-on='{ "require": ["window.googleMaps"], "run": "window.appContent.maps.load()", "data": "some-google-maps-key" }' />
```

I think you get the idea. This allowed us to tackle all the problems above in a simple _configuration_ based approach, and CSP was solved as well. 

We've been using this in 2sxc for Oqtane since the beginning, for example to run inline JS which a users App may have, since it can't run using any normal Oqtane mechanism.

2sxc also has an additional C# API so all our hybrid (Dnn/Oqtane) apps typically do something like 

```c#
Kit.Page.TurnOn("window.appContent.initFancybox()", data: data, require: "window.Fancybox");
```

With the latest JS challenges is Oqtane - one big one we haven't had time to explain here, but it's related to **Stream Rendering** which can have an html-document with a head, but no body for a short time, causing havoc on most JS loaded in the head... we decided to also use this to trigger the load js. 

Check out [the source code of blazor-cms.org](https://blazor-cms.org/). The initial source code has some `<turnOn>` tags. But now do a soft-load (enhanced) to another page, and look at the source in F12. You'll see that 2sxc adds code like this:

![image](https://github.com/user-attachments/assets/5d72d95c-4949-4217-b7a1-7e4f76c76085)

Note that it's a bit hard to comprehend what happens, because the tag which was added was much smaller, but when **turnOn** runs, it also updates the tag with information about what it checked, what happened, progress waiting for requirements and more. The real tag added was ca. 

```html
<turnoncjknwmojdnlc turn-on='{
  "run":"window.Oqtane.Interop.includeLinks()",
  "data":[ {
    "id":"",
    "rel":"stylesheet",
    "href":"/app/Content/assets/bs5/dist/styles.min.css",
    "type":"text/css",
    "integrity":"",
    "crossorigin":"",
    "insertbefore":"app-stylesheet-module"
}]}'/>
```

Note that we also were forced to randomize the tag name - here `<turnoncjknwmojdnlc ...>` because of the strange nature of the partial DOM updates. 

This is just an example. We do it a bit differently for Js, Css and inline JS, but from what we see it works just as expected in 99% of all cases. 

Note again that it _must_ be added to the body, as adding it to the head results in JS running at times where the `<body>` tag doesn't even exist, and most normal JS scripts will just break down if the `document` object and `window` object are in such a strange state.

**Comment by iJungleboy** [comment]
...if you feel like this would also work for Oqtane in general (I think it would) I would be glad to help and also enhance **turnOn** in case it would need some adjustments - eg. to `require` that the `html.body` exists before running scripts etc.

**Comment by sbwalker** [comment]
@iJungleboy I have reached out to Mackinnon Buck to explain the limitations of page-script which you described above. I am hoping he may be able to provide some guidance in this area. In general, the Oqtane philosophy is to be aligned with Microsoft rather than creating custom solutions which can cause potential maintenance and upgrade challenges in the future.

**Comment by iJungleboy** [comment]
@sbwalker makes sense - since anything we try to optimize may break as Microsoft figures out how to do things correctly as well.

An option is of course to try to provide temporary guidance - eg in the docs - which would not be part of Oqtane but help developers till there is a better solution. 

For now, let's wait for what Mackinnon Buck.
